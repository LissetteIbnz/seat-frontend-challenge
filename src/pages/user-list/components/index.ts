export * from "./user-row.component";
