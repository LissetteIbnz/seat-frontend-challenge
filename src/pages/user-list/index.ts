export * from "./user-list.container";
