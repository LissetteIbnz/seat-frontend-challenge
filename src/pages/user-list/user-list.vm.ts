import { Column } from "react-table";

export type UserID = number;
export type UserTitle = "Mr" | "Mrs" | "Dr" | "Honorable";

export interface User {
  id: UserID;
  fullName: string;
  email: string;
}

type KeysUser = keyof User;

export const FILTERS: KeysUser[] = ["fullName"];

export const HEADERS = [
  {
    Header: "Id",
    accessor: "id",
  },
  {
    Header: "Full Name",
    accessor: "fullName",
  },
  {
    Header: "Email",
    accessor: "email",
  },
  {
    Header: "",
    accessor: "actions",
    disableSortBy: true,
  },
] as Column<User>[];
