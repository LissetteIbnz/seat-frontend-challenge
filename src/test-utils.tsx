import * as React from "react";
import { fireEvent, render, RenderOptions, RenderResult } from "@testing-library/react";
import { ThemeProvider } from "./core/theme";
import { StylesProvider } from "@chakra-ui/react";

type RenderProps = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">,
) => RenderResult;

const WrapperProviders: React.FC = ({ children }) => (
  <StylesProvider value={{}}>
    <ThemeProvider>{children}</ThemeProvider>
  </StylesProvider>
);

const customRender: RenderProps = (ui, renderOptions) =>
  render(ui, { wrapper: WrapperProviders, ...renderOptions });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const changeInput = (input: HTMLInputElement | HTMLTextAreaElement, value: string) =>
  fireEvent.change(input, { target: { value } });
