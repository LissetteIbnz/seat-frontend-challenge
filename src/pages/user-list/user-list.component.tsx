import * as React from "react";
import { useSearchBar } from "common/components";
import { RowRendererProps, TableContainer } from "common/components/table";
import { UserRowComponent } from "./components";
import { FILTERS, HEADERS, User } from "./user-list.vm";

interface Props {
  onCreate: VoidFunction;
  onEdit: (id: string) => void;
  onDelete: (id: string) => void;
  users: User[];
}

export const UserListComponent = ({ users, onDelete, onCreate, onEdit }: Props) => {
  const { filteredList, onSearch, search } = useSearchBar(users, FILTERS);

  const contentRender = ({ itemName }: { itemName: string }) => {
    return <>Are you sure to delete {itemName}? You can't undo this action afterwards.</>;
  };

  const labels = React.useMemo(
    () => ({
      acceptButton: "Confirm",
      closeButton: "Cancel",
      createButton: "New User",
      deleteContent: contentRender,
      deleteTitle: "Delete User",
      searchPlaceholder: "Search user",
    }),
    [],
  );

  return (
    <TableContainer<User>
      onCreate={onCreate}
      onEdit={onEdit}
      columns={HEADERS}
      enablePagination={true}
      enableSearch={true}
      labels={labels}
      onDelete={onDelete}
      onSearch={onSearch}
      pageSize={5}
      rowRenderer={(rowProps: RowRendererProps<User>) => <UserRowComponent {...rowProps} />}
      rows={filteredList}
      search={search}
      sx={{
        maxWidth: "1170px",
        width: "100%",
      }}
    />
  );
};
