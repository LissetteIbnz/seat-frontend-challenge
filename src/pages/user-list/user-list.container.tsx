import * as React from "react";
import { useNavigate } from "react-router-dom";
import { useToast } from "@chakra-ui/react";
import { routes } from "core/router";
import { getErrorMessage } from "common/helpers";
import { UserListComponent } from "./user-list.component";
import { User } from "./user-list.vm";
import { deleteUser, getAllUsers } from "./user-list.service";

// TODO: Pending to implement an error handler and loader component (handle no results found flow).
export const UserListContainer = () => {
  const navigate = useNavigate();
  const toast = useToast({
    position: "top-right",
  });

  const [users, setUsers] = React.useState<User[]>([]);
  const [isLoading, setIsLoading] = React.useState(false);

  const onLoadUsers = async () => {
    setIsLoading(true);
    try {
      const usersList = await getAllUsers();
      setUsers(usersList);
    } catch (error) {
      const message = getErrorMessage(error);
      toast({
        description: message,
        status: "error",
      });
    }
    setIsLoading(false);
  };

  React.useEffect(() => {
    onLoadUsers();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const handleDelete = async (userId: string) => {
    try {
      await deleteUser(userId);
      setUsers(users.filter((user) => user.id.toString() !== userId));
      toast({
        description: "User deleted successfully",
        status: "success",
      });
    } catch (error) {
      toast({
        description: getErrorMessage(error),
        status: "error",
      });
    }
  };

  const handleCreate = () => {
    navigate(routes.createUser);
  };

  const handleEdit = (id: string) => {
    navigate(routes.editUser(id));
  };

  if (isLoading) {
    return <div>Loading...</div>;
  }

  return (
    <UserListComponent
      onCreate={handleCreate}
      onDelete={handleDelete}
      onEdit={handleEdit}
      users={users}
    />
  );
};
