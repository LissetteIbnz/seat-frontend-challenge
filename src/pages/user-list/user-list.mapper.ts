import { mapToCollection } from "common/mappers";
import * as AM from "infra/dto";
import * as VM from "./user-list.vm";

export const mapUserFromApiToVm = (user: AM.UserDTO): VM.User => ({
  id: user.id,
  fullName: [user.first_name, user.last_name].join(" ").trim(),
  email: user.email,
});

export const mapUserListFromApiToVm = (users: AM.UserDTO[]): VM.User[] =>
  mapToCollection(users, mapUserFromApiToVm);
