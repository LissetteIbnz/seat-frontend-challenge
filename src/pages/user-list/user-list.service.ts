import { userRepository } from "infra/repositories";
import { mapUserListFromApiToVm } from "./user-list.mapper";
import * as VM from "./user-list.vm";

export const getAllUsers = async (): Promise<VM.User[]> => {
  const apiUsersList = await userRepository.getAll();
  const vmUsersList = mapUserListFromApiToVm(apiUsersList);
  return vmUsersList;
};

export const deleteUser = async (id: string): Promise<void> => {
  await userRepository.delete(id);
};
