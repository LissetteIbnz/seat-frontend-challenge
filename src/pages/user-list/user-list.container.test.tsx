import { toast } from "@chakra-ui/react";
import { MemoryRouter } from "react-router-dom";
import { render, screen, fireEvent, waitFor } from "test-utils";
import { UserListContainer } from "./user-list.container";
import * as UserService from "./user-list.service";

const mockedUseNavigate = jest.fn();

jest.mock("./user-list.service");
jest.mock("react-router-dom", () => {
  const originalModule = jest.requireActual("react-router-dom");
  return {
    ...originalModule,
    useNavigate: () => mockedUseNavigate,
  };
});

const renderWithRouter = () => {
  return render(
    <MemoryRouter>
      <UserListContainer />
    </MemoryRouter>,
  );
};

describe("pages/user-list/UserListContainer", () => {
  // ðŸ‘‡ Close all toasts before each tests and wait for them to be removed.
  beforeEach(() => {
    toast.closeAll();
    document.getElementById("chakra-toast-portal")?.remove();
  });

  it("should render correctly", async () => {
    jest.spyOn(UserService, "getAllUsers").mockResolvedValueOnce([]);

    const { container } = renderWithRouter();

    await waitFor(() => {
      expect(screen.queryByText(/loading/i)).not.toBeInTheDocument();
    });
    expect(container).toMatchSnapshot();
  });

  it("should render the UserListComponent", async () => {
    jest.spyOn(UserService, "getAllUsers").mockResolvedValueOnce([
      {
        id: 1,
        fullName: "John Doe",
        email: "irrelevant",
      },
    ]);

    renderWithRouter();

    await waitFor(() => {
      expect(screen.queryByText(/loading/i)).not.toBeInTheDocument();
    });

    expect(screen.getByText(/John Doe/i)).toBeInTheDocument();
  });

  it("should handle to create a user", async () => {
    jest.spyOn(UserService, "getAllUsers").mockResolvedValueOnce([]);

    renderWithRouter();

    await waitFor(() => {
      expect(screen.queryByText(/loading/i)).not.toBeInTheDocument();
    });

    const createButton = screen.getByRole("button", { name: /new user/i });
    fireEvent.click(createButton);

    expect(mockedUseNavigate).toHaveBeenCalledTimes(1);
  });

  it("should handle to edit a user", async () => {
    jest.spyOn(UserService, "getAllUsers").mockResolvedValueOnce([
      {
        id: 1,
        fullName: "John Doe",
        email: "irrelevant",
      },
    ]);

    renderWithRouter();

    await waitFor(() => {
      expect(screen.queryByText(/loading/i)).not.toBeInTheDocument();
    });

    const editButton = screen.getByRole("button", { name: /edit/i });
    fireEvent.click(editButton);

    expect(mockedUseNavigate).toHaveBeenCalledTimes(1);
  });

  it("should handle to delete a user", async () => {
    jest.spyOn(UserService, "getAllUsers").mockResolvedValueOnce([
      {
        id: 1,
        fullName: "John Doe",
        email: "irrelevant",
      },
    ]);

    jest.spyOn(UserService, "deleteUser").mockResolvedValueOnce();

    renderWithRouter();

    await waitFor(() => {
      expect(screen.queryByText(/loading/i)).not.toBeInTheDocument();
    });

    const deleteButton = screen.getByRole("button", { name: /delete/i });
    fireEvent.click(deleteButton);

    const confirmButton = screen.getByRole("button", { name: /confirm/i });
    fireEvent.click(confirmButton);

    await waitFor(() => {
      expect(screen.getByText(/User deleted successfully/i)).toBeInTheDocument();
    });
  });

  it("should handle to cancel delete a user", async () => {
    jest.spyOn(UserService, "getAllUsers").mockResolvedValueOnce([
      {
        id: 1,
        fullName: "John Doe",
        email: "irrelevant",
      },
    ]);

    const spyDeleteUser = jest.spyOn(UserService, "deleteUser").mockResolvedValueOnce();

    renderWithRouter();

    await waitFor(() => {
      expect(screen.queryByText(/loading/i)).not.toBeInTheDocument();
    });

    const deleteButton = screen.getByRole("button", { name: /delete/i });
    fireEvent.click(deleteButton);

    const cancelButton = screen.getByRole("button", { name: /cancel/i });
    fireEvent.click(cancelButton);

    expect(spyDeleteUser).not.toHaveBeenCalled();
  });
});
