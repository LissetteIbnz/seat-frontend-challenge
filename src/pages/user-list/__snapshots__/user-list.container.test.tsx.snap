// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/user-list/UserListContainer should render correctly 1`] = `
<div>
  <div
    class="css-1kl3tk9"
  >
    <div
      class="css-j7qwjs"
    >
      <div
        class="chakra-input__group css-1hm0tut"
      >
        <div
          class="chakra-input__left-element css-1cw84h2"
        >
          <svg
            fill="currentColor"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
            x="0px"
            xmlns="http://www.w3.org/2000/svg"
            y="0px"
          >
            <path
              d="M 9 2 C 5.1458514 2 2 5.1458514 2 9 C 2 12.854149 5.1458514 16 9 16 C 10.747998 16 12.345009 15.348024 13.574219 14.28125 L 14 14.707031 L 14 16 L 20 22 L 22 20 L 16 14 L 14.707031 14 L 14.28125 13.574219 C 15.348024 12.345009 16 10.747998 16 9 C 16 5.1458514 12.854149 2 9 2 z M 9 4 C 11.773268 4 14 6.2267316 14 9 C 14 11.773268 11.773268 14 9 14 C 6.2267316 14 4 11.773268 4 9 C 4 6.2267316 6.2267316 4 9 4 z"
            />
          </svg>
        </div>
        <input
          class="chakra-input css-8tczqk"
          placeholder="Search user"
          type="search"
          value=""
        />
      </div>
      <div
        class="chakra-table__container css-zipzvv"
      >
        <table
          class="chakra-table css-1kvgk9m"
          role="table"
        >
          <thead
            class="css-1h4etlm"
          >
            <tr
              class="css-0"
              role="row"
            >
              <th
                class="css-1lpxuf2"
                colspan="1"
                role="columnheader"
                style="cursor: pointer;"
                title="Toggle SortBy"
              >
                Id
                <span
                  class="css-176c8l1"
                />
              </th>
              <th
                class="css-1lpxuf2"
                colspan="1"
                role="columnheader"
                style="cursor: pointer;"
                title="Toggle SortBy"
              >
                Full Name
                <span
                  class="css-176c8l1"
                />
              </th>
              <th
                class="css-1lpxuf2"
                colspan="1"
                role="columnheader"
                style="cursor: pointer;"
                title="Toggle SortBy"
              >
                Email
                <span
                  class="css-176c8l1"
                />
              </th>
              <th
                class="css-1lpxuf2"
                colspan="1"
                role="columnheader"
              >
                
                <span
                  class="css-176c8l1"
                />
              </th>
            </tr>
          </thead>
          <tbody
            class="css-tdnrhj"
          >
            <tr
              class="css-0"
              role="row"
            >
              <td
                class="css-xumdn4"
                colspan="1"
                role="gridcell"
              >
                No search results found
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <button
        class="chakra-button css-1bxdac1"
        type="button"
      >
        New User
      </button>
    </div>
  </div>
</div>
`;
